src/core/src/bound_evaluate.cpp

```diff
@@ -75,85 +75,219 @@ bool is_type_allocable(const element::Type& type) {
     return type != element::undefined && type.is_static();
 }
 
-ov::Tensor evaluate_bound(const Output<Node>& output, bool is_upper, bool invalidate_all_unused_values = true) {
-    if (is_upper && output.get_tensor().get_upper_value()) {
-        return output.get_tensor().get_upper_value();
-    }
+/**
+ * @brief Checks if node's inputs has bound set and are equal.
+ *
+ * @param node  Node
+ * @return true if all inputs have both bounds set and are equal, otherwise false.
+ */
+bool node_has_same_inputs(const Node& node) {
+    const auto& values = node.input_values();
+
+    return std::all_of(values.begin(), values.end(), [](const Output<Node>& value) {
+        const auto& t_desc = value.get_tensor();
+        return t_desc.has_and_set_bound() || are_equal(t_desc.get_lower_value(), t_desc.get_upper_value());
```

Comment:
Should we also check `are_same_tensor` here as a shortcut?
Additionally, since all the helpers are checking it, `t_desc.has_and_set_bound()` seems duplicated.